tools/fpgabist/bist/bist_afu.c

```diff
@@ -0,0 +1,369 @@
+	}
+
+	/* Reset accelerator */
+	res = fpgaReset(accelerator_handle);
+	ON_ERR_GOTO(res, out_free_output, "resetting accelerator");
+
+	/* Program DMA addresses */
+	uint64_t iova;
+	res = fpgaGetIOAddress(accelerator_handle, dsm_wsid, &iova);
+	ON_ERR_GOTO(res, out_free_output, "getting DSM IOVA");
+
+	res = fpgaWriteMMIO64(accelerator_handle, 0, CSR_AFU_DSM_BASEL, iova);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_AFU_DSM_BASEL");
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CTL, 0);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CTL, 1);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	res = fpgaGetIOAddress(accelerator_handle, input_wsid, &iova);
+	ON_ERR_GOTO(res, out_free_output, "getting input IOVA");
+	res = fpgaWriteMMIO64(accelerator_handle, 0, CSR_SRC_ADDR, CACHELINE_ALIGNED_ADDR(iova));
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_SRC_ADDR");
+
+	res = fpgaGetIOAddress(accelerator_handle, output_wsid, &iova);
+	ON_ERR_GOTO(res, out_free_output, "getting output IOVA");
+	res = fpgaWriteMMIO64(accelerator_handle, 0, CSR_DST_ADDR, CACHELINE_ALIGNED_ADDR(iova));
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_DST_ADDR");
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_NUM_LINES, LPBK1_BUFFER_SIZE / CL(1));
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_NUM_LINES");
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CFG, 0x42000);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	status_ptr = dsm_ptr + DSM_STATUS_TEST_COMPLETE/8;
+
+	/* Start the test */
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CTL, 3);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	/* Wait for test completion */
+	while (0 == ((*status_ptr) & 0x1)) {
+		usleep(100);
+	}
+
+
+/************* Begin BIST *************/
+        /* Perform BIST Check */
+	printf("Running BIST Test\n");
+        uint64_t data = 0;
+        double count = 0;
+        unsigned int bist_mask = ENABLE_DDRA_BIST;
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
+	ON_ERR_GOTO(res, out_free_output, "write DDR_BIST_CTRL_ADDR");
+
+        res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_CTRL_ADDR,&data);
+	ON_ERR_GOTO(res, out_free_output, "write DDR_BIST_CTRL_ADDR");
+
+	while((CHECK_BIT(data,27) != 0x08000000)){
+	  printf("Enable Test: reading result #%f: %04lx\n", count, data);
+	  if (count >= MAX_COUNT){
+		fprintf(stderr, "BIST not enabled!\n");
+                return -1;
+          }
+	  count++;
+	}
+	printf("BIST is enabled.  Reading status register\n");
+
+	count = 0;
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_BIST");
+        while ((CHECK_BIT(data,9) != 0x200) && (CHECK_BIT(data,8) != 0x100) && (CHECK_BIT(data,7) != 0x80) && 
+            (CHECK_BIT(data,10) != 0x400) && (CHECK_BIT(data,11) != 0x800) && (CHECK_BIT(data,12) != 0x1000)){
+          res = fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_STATUS_ADDR,
```

Comment:
Can this `while` condition be simplified? Maybe in a follow up PR?

Reply:
future changes like this are going to be made in future PRs

Reply:
Right, that's why I say "in a follow up PR". How are you tracking these TODO things?